Network/DNS/Resolver.hs

```diff
@@ -397,23 +383,30 @@ lookupRawInternal ::
     -> IO (Either DNSError DNSMessage)
 lookupRawInternal _ _ _   dom _
   | isIllegal dom     = return $ Left IllegalDomain
-lookupRawInternal rcv ad rlv dom typ = loop (NE.uncons (dnsSocks rlv))
+lookupRawInternal rcv ad rlv dom typ = loop (NE.uncons (dnsServers rlv))
   where
-    loop :: (Socket, Maybe (NonEmpty Socket)) -> IO (Either DNSError DNSMessage)
-    loop (sock, alternatives) = do
+    loop :: (AddrInfo, Maybe (NonEmpty AddrInfo)) -> IO (Either DNSError DNSMessage)
+    loop (ai, ais) = do
       res <- initialise >>= \(query, checkSeqno) ->
-        performLookup sock query checkSeqno 0 False
+        bracket (udpOpen ai)
+                (close)
+                (\sock -> performLookup ai sock query checkSeqno 0 False)
       case res of
-        Left e  -> maybe (return (Left e)) (loop . NE.uncons) alternatives
+        Left e  -> maybe (return (Left e)) (loop . NE.uncons) ais
         Right v -> pure (Right v)
 
+    udpOpen ai = do
+        sock <- socket (addrFamily ai) (addrSocketType ai) (addrProtocol ai)
+        connect sock (addrAddress ai)
```

Comment:
There's a small chance of raising an uncaught I/O error here, if socket() fails due to file-descriptor exhaustion (system-wide or per-process limits). Same issue applied to `withResover` in the original code, we could attempt to address this.  Likewise the connect() could fail due to failure acquire a free local port, and then we also leak the socket.  Both conditions are rather uncommon.  Perhaps just a comment for now, and some appropriate error handling later. We could return "NetworkError" or create a new branch of the DNSError sum type for SystemError (these are local resource problems).

Reply:
> Perhaps just a comment for now, and some appropriate error handling later. 

Agree.

Reply:
Comments for udpOpen and tcpOpen fixed in next commit.